File: worker/app/services/summary_audio_player.py
import base64
import os

from app.services.audio_briefing_tts import synthesize_mock_audio
from app.services.aivis_speech import AivisSpeechService
from app.services.tts_provider_metadata import _env_float
from app.services.tts_provider_metadata import load_single_speaker_tts_provider_runtime_metadata
from app.services.tts_provider_registry import synthesize_single_speaker_tts


def build_summary_audio_text(translated_title: str | None, original_title: str | None, summary: str) -> str:
    title = (translated_title or "").strip() or (original_title or "").strip()
    summary_text = (summary or "").strip()
    if not title:
        raise RuntimeError("summary audio title is empty")
    if not summary_text:
        raise RuntimeError("summary audio summary is empty")
    return f"{title}\n\n{summary_text}"


class SummaryAudioPlayerService:
    def __init__(self) -> None:
        self.aivis = AivisSpeechService()
        self.single_speaker_provider_runtime = {
            provider: load_single_speaker_tts_provider_runtime_metadata(provider)
            for provider in ("xai", "gemini_tts", "fish", "elevenlabs", "openai", "azure_speech")
        }
        self.gemini_tts_endpoint = self.single_speaker_provider_runtime["gemini_tts"].endpoint
        self.gemini_timeout_sec = self.single_speaker_provider_runtime["gemini_tts"].timeout_sec
        self.fish_api_key = self.single_speaker_provider_runtime["fish"].api_key
        self.fish_timeout_sec = self.single_speaker_provider_runtime["fish"].timeout_sec
        self.elevenlabs_tts_endpoint = self.single_speaker_provider_runtime["elevenlabs"].endpoint
        self.elevenlabs_api_key = self.single_speaker_provider_runtime["elevenlabs"].api_key
        self.elevenlabs_timeout_sec = self.single_speaker_provider_runtime["elevenlabs"].timeout_sec

    def synthesize(
        self,
        *,
        provider: str,
        voice_model: str,
        voice_style: str,
        text: str,
        speech_rate: float,
        emotional_intensity: float,
        tempo_dynamics: float,
        line_break_silence_seconds: float,
        chunk_trailing_silence_seconds: float,
        pitch: float,
        volume_gain: float,
        tts_model: str = "",
        user_dictionary_uuid: str | None = None,
        aivis_api_key: str | None = None,
        google_api_key: str | None = None,
        fish_api_key: str | None = None,
        elevenlabs_api_key: str | None = None,
        xai_api_key: str | None = None,
        openai_api_key: str | None = None,
        azure_speech_api_key: str | None = None,
        azure_speech_region: str | None = None,
    ) -> tuple[str, str, int, str]:
        normalized_provider = (provider or "").strip().lower()
        if normalized_provider == "mock":
            audio_bytes, content_type, _, duration_sec = synthesize_mock_audio(text, speech_rate)
        elif normalized_provider == "aivis":
            audio_bytes, content_type, _, duration_sec = self.aivis.synthesize(
                voice_model=voice_model,
                voice_style=voice_style,
                text=text,
                speech_rate=speech_rate,
                emotional_intensity=emotional_intensity,
                tempo_dynamics=tempo_dynamics,
                line_break_silence_seconds=line_break_silence_seconds,
                chunk_trailing_silence_seconds=chunk_trailing_silence_seconds,
                pitch=pitch,
                volume_gain=volume_gain,
                user_dictionary_uuid=user_dictionary_uuid,
                api_key_override=aivis_api_key,
            )
        elif normalized_provider in {"xai", "gemini_tts", "fish", "elevenlabs", "openai", "azure_speech"}:
            runtime = self.single_speaker_provider_runtime[normalized_provider]
            api_key = runtime.api_key
            region = runtime.region
            if normalized_provider == "xai":
                api_key = (xai_api_key or "").strip() or api_key
            elif normalized_provider == "gemini_tts":
                api_key = (google_api_key or "").strip() or api_key
            elif normalized_provider == "fish":
                api_key = (fish_api_key or "").strip() or api_key
            elif normalized_provider == "elevenlabs":
                api_key = (elevenlabs_api_key or "").strip() or api_key
            elif normalized_provider == "openai":
                api_key = (openai_api_key or "").strip() or api_key
            elif normalized_provider == "azure_speech":
                api_key = (azure_speech_api_key or "").strip() or api_key
                region = (azure_speech_region or "").strip() or region
            audio_bytes, content_type, _, duration_sec = synthesize_single_speaker_tts(
                normalized_provider,
                endpoint=runtime.endpoint,
                api_key=api_key,
                region=region,
                voice_id=voice_model,
                tts_model=tts_model,
                text=text,
                speech_rate=speech_rate,
                timeout_sec=runtime.timeout_sec,
                volume_gain=volume_gain,
                line_break_silence_seconds=line_break_silence_seconds,
                pitch=pitch,
            )
        else:
            raise RuntimeError(f"unsupported summary audio provider: {provider}")
        return base64.b64encode(audio_bytes).decode("utf-8"), content_type, duration_sec, text

File: api/internal/service/summary_audio_player.go
package service

import (
	"context"
	"errors"
	"strings"

	"github.com/enjoydarts/sifto/api/internal/model"
	"github.com/enjoydarts/sifto/api/internal/repository"
)

var (
	ErrSummaryAudioMissingSummary = errors.New("summary audio summary is not available")
	ErrSummaryAudioMissingVoice   = errors.New("summary audio voice is not configured")
	ErrSummaryAudioMissingModel   = errors.New("summary audio model is not configured")
)

type SummaryAudioSynthesis struct {
	Item             *model.ItemDetail `json:"item,omitempty"`
	Persona          string            `json:"persona"`
	AudioBase64      string            `json:"audio_base64"`
	ContentType      string            `json:"content_type"`
	DurationSec      int               `json:"duration_sec"`
	ResolvedText     string            `json:"resolved_text"`
	PreprocessedText *string           `json:"preprocessed_text,omitempty"`
}

type SummaryAudioPlayerService struct {
	items        *repository.ItemRepo
	summaryAudio *repository.SummaryAudioVoiceSettingsRepo
	userRepo     *repository.UserRepo
	userSettings *repository.UserSettingsRepo
	cipher       *SecretCipher
	worker       summaryAudioSynthesizer
	preprocess   summaryAudioTTSMarkupPreprocessor
}

type summaryAudioSynthesizer interface {
	SynthesizeSummaryAudio(
		ctx context.Context,
		provider string,
		voiceModel string,
		voiceStyle string,
		ttsModel string,
		azureSpeechRegion string,
		text string,
		speechRate float64,
		emotionalIntensity float64,
		tempoDynamics float64,
		lineBreakSilenceSeconds float64,
		chunkTrailingSilenceSeconds float64,
		pitch float64,
		volumeGain float64,
		aivisUserDictionaryUUID *string,
		aivisAPIKey *string,
		fishAudioAPIKey *string,
		elevenLabsAPIKey *string,
		googleAPIKey *string,
		xaiAPIKey *string,
		openAIAPIKey *string,
		azureSpeechAPIKey *string,
	) (*SummaryAudioSynthesizeResponse, error)
}

type summaryAudioTTSMarkupPreprocessor interface {
	PreprocessSummaryAudioText(ctx context.Context, userID, itemID, text string) (*TTSMarkupPreprocessResult, error)
	PreprocessSummaryAudioTextForProvider(ctx context.Context, userID, itemID, provider, text string) (*TTSMarkupPreprocessResult, error)
	PreprocessSummaryAudioTextForProviderWithVariables(ctx context.Context, userID, itemID, provider, text string, variables map[string]string) (*TTSMarkupPreprocessResult, error)
}

func NewSummaryAudioPlayerService(
	items *repository.ItemRepo,
	summaryAudio *repository.SummaryAudioVoiceSettingsRepo,
	userRepo *repository.UserRepo,
	userSettings *repository.UserSettingsRepo,
	cipher *SecretCipher,
	worker summaryAudioSynthesizer,
	preprocess summaryAudioTTSMarkupPreprocessor,
) *SummaryAudioPlayerService {
	return &SummaryAudioPlayerService{
		items:        items,
		summaryAudio: summaryAudio,
		userRepo:     userRepo,
		userSettings: userSettings,
		cipher:       cipher,
		worker:       worker,
		preprocess:   preprocess,
	}
}

func BuildSummaryAudioNarration(translatedTitle, originalTitle, summary string) string {
	title := strings.TrimSpace(translatedTitle)
	if title == "" {
		title = strings.TrimSpace(originalTitle)
	}
	return title + "\n\n" + strings.TrimSpace(summary)
}

func SummaryAudioRequestContext(parent context.Context) context.Context {
	if parent == nil {
		return context.Background()
	}
	return context.WithoutCancel(parent)
}

func (s *SummaryAudioPlayerService) Synthesize(ctx context.Context, userID, itemID string) (*SummaryAudioSynthesis, error) {
	if s == nil || s.items == nil || s.summaryAudio == nil || s.userSettings == nil || s.worker == nil {
		return nil, errors.New("summary audio service is not configured")
	}
	item, err := s.items.GetDetail(ctx, strings.TrimSpace(itemID), strings.TrimSpace(userID))
	if err != nil {
		return nil, err
	}
	summaryText := summaryAudioSummaryText(item)
	if summaryText == "" {
		return nil, ErrSummaryAudioMissingSummary
	}
	settings, err := s.summaryAudio.EnsureDefaults(ctx, userID)
	if err != nil {
		return nil, err
	}
	provider := strings.TrimSpace(settings.TTSProvider)
	voiceModel := strings.TrimSpace(settings.VoiceModel)
	voiceStyle := strings.TrimSpace(settings.VoiceStyle)
	ttsModel := strings.TrimSpace(settings.TTSModel)
	providerMetadata := LookupTTSProviderMetadata(provider)
	if provider == "" || voiceModel == "" {
		return nil, ErrSummaryAudioMissingVoice
	}
	narration := BuildSummaryAudioNarration(derefString(item.TranslatedTitle), derefString(item.Title), summaryText)
	var preprocessedText *string
	var aivisAPIKey *string
	var aivisUserDictionaryUUID *string
	var fishAudioAPIKey *string
	var elevenLabsAPIKey *string
	var googleAPIKey *string
	var xaiAPIKey *string
	var openAIAPIKey *string
	var azureSpeechAPIKey *string
	azureSpeechRegion := ""
	if strings.EqualFold(provider, "aivis") {
		aivisAPIKey, err = s.loadAivisAPIKey(ctx, userID)
		if err != nil {
			return nil, err
		}
		aivisUserDictionaryUUID = settings.AivisUserDictionaryUUID
	} else if strings.EqualFold(provider, "xai") {
		xaiAPIKey, err = s.loadXAIAPIKey(ctx, userID)
		if err != nil {
			return nil, err
		}
		if s.preprocess != nil {
			preprocessed, preprocessErr := s.preprocess.PreprocessSummaryAudioTextForProvider(ctx, userID, item.ID, provider, narration)
			if preprocessErr != nil {
				return nil, preprocessErr
			}
			preprocessedText = stringPtrOrNil(preprocessed.Text)
			narration = preprocessed.Text
		}
	} else if strings.EqualFold(provider, "fish") {
		if providerMetadata.SummaryRequiresTTSModel && ttsModel == "" {
			return nil, ErrSummaryAudioMissingModel
		}
		fishAudioAPIKey, err = loadAndDecryptAudioBriefingUserSecret(ctx, s.userSettings.GetFishAudioAPIKeyEncrypted, s.cipher, userID, "fish api key is not configured")
		if err != nil {
			return nil, err
		}
		if s.preprocess != nil {
			preprocessed, preprocessErr := s.preprocess.PreprocessSummaryAudioTextForProvider(ctx, userID, item.ID, provider, narration)
			if preprocessErr != nil {
				return nil, preprocessErr
			}
			preprocessedText = stringPtrOrNil(preprocessed.Text)
			narration = preprocessed.Text
		}
	} else if strings.EqualFold(provider, "gemini_tts") {
		if providerMetadata.SummaryRequiresTTSModel && ttsModel == "" {
			return nil, ErrSummaryAudioMissingModel
		}
		if err := EnsureGeminiTTSEnabledForUser(ctx, s.userRepo, userID); err != nil {
			return nil, err
		}
		if s.preprocess != nil {
			preprocessed, preprocessErr := s.preprocess.PreprocessSummaryAudioTextForProvider(ctx, userID, item.ID, provider, narration)
			if preprocessErr != nil {
				return nil, preprocessErr
			}
			preprocessedText = stringPtrOrNil(preprocessed.Text)
			narration = preprocessed.Text
		}
	} else if strings.EqualFold(provider, "elevenlabs") {
		if providerMetadata.SummaryRequiresTTSModel && ttsModel == "" {
			return nil, ErrSummaryAudioMissingModel
		}
		elevenLabsAPIKey, err = loadAndDecryptAudioBriefingUserSecret(ctx, s.userSettings.GetElevenLabsAPIKeyEncrypted, s.cipher, userID, "elevenlabs api key is not configured")
		if err != nil {
			return nil, err
		}
		if s.preprocess != nil {
			preprocessed, preprocessErr := s.preprocess.PreprocessSummaryAudioTextForProvider(ctx, userID, item.ID, provider, narration)
			if preprocessErr != nil {
				return nil, preprocessErr
			}
			preprocessedText = stringPtrOrNil(preprocessed.Text)
			narration = preprocessed.Text
		}
	} else if strings.EqualFold(provider, "openai") {
		if providerMetadata.SummaryRequiresTTSModel && ttsModel == "" {
			return nil, ErrSummaryAudioMissingModel
		}
		openAIAPIKey, err = loadAndDecryptAudioBriefingUserSecret(ctx, s.userSettings.GetOpenAIAPIKeyEncrypted, s.cipher, userID, "openai api key is not configured")
		if err != nil {
			return nil, err
		}
	} else if strings.EqualFold(provider, "azure_speech") {
		azureSpeechAPIKey, err = loadAndDecryptAudioBriefingUserSecret(ctx, s.userSettings.GetAzureSpeechAPIKeyEncrypted, s.cipher, userID, "azure speech api key is not configured")
		if err != nil {
			return nil, err
		}
		region, regionErr := s.userSettings.GetAzureSpeechRegion(ctx, userID)
		if regionErr != nil {
			return nil, regionErr
		}
		azureSpeechRegion = strings.TrimSpace(derefString(region))
		if azureSpeechRegion == "" {
			return nil, errors.New("azure speech region is not configured")
		}
		if s.preprocess != nil {
			preprocessed, preprocessErr := s.preprocess.PreprocessSummaryAudioTextForProviderWithVariables(ctx, userID, item.ID, provider, narration, map[string]string{
				"voice_name":   strings.TrimSpace(voiceModel),
				"voice_locale": "ja-JP",
			})
			if preprocessErr != nil {
				return nil, preprocessErr
			}
			preprocessedText = stringPtrOrNil(preprocessed.Text)
			narration = preprocessed.Text
		}
	}
	resp, err := s.worker.SynthesizeSummaryAudio(
		ctx,
		provider,
		voiceModel,
		voiceStyle,
		ttsModel,
		azureSpeechRegion,
		narration,
		settings.SpeechRate,
		settings.EmotionalIntensity,
		settings.TempoDynamics,
		settings.LineBreakSilenceSeconds,
		1.0,
		settings.Pitch,
		settings.VolumeGain,
		aivisUserDictionaryUUID,
		aivisAPIKey,
		fishAudioAPIKey,
		elevenLabsAPIKey,
		googleAPIKey,
		xaiAPIKey,
		openAIAPIKey,
		azureSpeechAPIKey,
	)
	if err != nil {
		return nil, err
	}
	return &SummaryAudioSynthesis{
		Item:             item,
		Persona:          "",
		AudioBase64:      resp.AudioBase64,
		ContentType:      resp.ContentType,
		DurationSec:      resp.DurationSec,
		ResolvedText:     resp.ResolvedText,
		PreprocessedText: preprocessedText,
	}, nil
}

func (s *SummaryAudioPlayerService) loadXAIAPIKey(ctx context.Context, userID string) (*string, error) {
	if s == nil || s.userSettings == nil || s.cipher == nil {
		return nil, errors.New("summary audio xai key loader is not configured")
	}
	enc, err := s.userSettings.GetXAIAPIKeyEncrypted(ctx, userID)
	if err != nil {
		return nil, err
	}
	if enc == nil || strings.TrimSpace(*enc) == "" {
		return nil, errors.New("xai api key is not configured")
	}
	plain, err := s.cipher.DecryptString(*enc)
	if err != nil {
		return nil, err
	}
	plain = strings.TrimSpace(plain)
	if plain == "" {
		return nil, errors.New("xai api key is empty")
	}
	return &plain, nil
}

func (s *SummaryAudioPlayerService) loadAivisAPIKey(ctx context.Context, userID string) (*string, error) {
	if s == nil || s.userSettings == nil || s.cipher == nil {
		return nil, errors.New("summary audio aivis key loader is not configured")
	}
	enc, err := s.userSettings.GetAivisAPIKeyEncrypted(ctx, userID)
	if err != nil {
		return nil, err
	}
	if enc == nil || strings.TrimSpace(*enc) == "" {
		return nil, errors.New("aivis api key is not configured")
	}
	plain, err := s.cipher.DecryptString(*enc)
	if err != nil {
		return nil, err
	}
	plain = strings.TrimSpace(plain)
	if plain == "" {
		return nil, errors.New("aivis api key is empty")
	}
	return &plain, nil
}

func summaryAudioSummaryText(item *model.ItemDetail) string {
	if item == nil {
		return ""
	}
	if item.Summary != nil {
		if text := strings.TrimSpace(item.Summary.Summary); text != "" {
			return text
		}
	}
	return strings.TrimSpace(derefString(item.Item.Summary))
}
